lib/datadog/di/probe_notification_builder.rb

```diff
@@ -0,0 +1,212 @@
+                arguments: if serialized_entry_args
+                  serialized_entry_args
+                else
+                  (args || kwargs) && serializer.serialize_args(args, kwargs)
+                end,
+                throwable: nil,
+                # standard:enable all
+              },
+              return: {
+                arguments: {
+                  "@return": serializer.serialize_value(rv),
+                },
+                throwable: nil,
+              },
+            }
+          elsif probe.line?
+            {
+              lines: snapshot && {
+                probe.line_no => {locals: serializer.serialize_vars(snapshot)},
+              },
+            }
+          end
+        end
+
+        location = if probe.line?
+          actual_file = if probe.file
+            # Normally caller_locations should always be filled for a line probe
+            # but in the test suite we don't always provide all arguments.
+            actual_file_basename = File.basename(probe.file)
+            caller_locations&.detect do |loc|
+              # TODO record actual path that probe was installed into,
+              # perform exact match here against that path.
+              File.basename(loc.path) == actual_file_basename
+            end&.path || probe.file
+          end
+          {
+            file: actual_file,
+            lines: [probe.line_no],
+          }
+        elsif probe.method?
+          {
+            method: probe.method_name,
+            type: probe.type_name,
+          }
+        end
+
+        stack = if caller_locations
+          format_caller_locations(caller_locations)
+        end
+
+        timestamp = timestamp_now
+        {
+          service: settings.service,
+          "debugger.snapshot": {
+            id: SecureRandom.uuid,
+            timestamp: timestamp,
+            evaluationErrors: [],
+            probe: {
+              id: probe.id,
+              version: 0,
+              location: location,
+            },
+            language: 'ruby',
+            # TODO add test coverage for callers being nil
+            stack: stack,
+            captures: captures,
+          },
+          # In python tracer duration is under debugger.snapshot,
+          # but UI appears to expect it here at top level.
+          duration: duration ? (duration * 10**9).to_i : nil,
+          host: nil,
+          logger: {
+            name: probe.file,
+            method: probe.method_name || 'no_method',
+            thread_name: Thread.current.name,
+            thread_id: thread_id,
+            version: 2,
+          },
+          # TODO add tests that the trace/span id is correctly propagated
+          "dd.trace_id": Datadog::Tracing.active_trace&.id,
+          "dd.span_id": Datadog::Tracing.active_span&.id,
+          ddsource: 'dd_debugger',
+          message: probe.template && evaluate_template(probe.template,
+            duration: duration ? duration * 1000 : nil),
```

Comment:
Minor: It may be worth encoding the units of `duration` somewhere in the variable names (`duration_ms`?) or using an explicit type, since units are getting mixed and converted a bit and it seems easy to make a mistake...

Reply:
The convention I use is when units are not specified, time is measured in seconds. In this particular case, DI specifies that  duration in the template should be measured in milliseconds, but the variable is called `duration`. Thus I am not sure how to add a suffix here.

Reply:
Even if the key itself in the data is duration, maybe our variable can be `duration_s` or `duration_secs`?

Reply:
I added a comment stating the units of duration for the method parameter. I personally have the convention that time units without an explicitly stated unit are in seconds, which I believe is how everything in Ruby standard library works, and to me adding a suffix for seconds is redundant. But if you want a suffix let's go with whichever one you prefer.

Reply:
I totally agree with @ivoanjo regarding naming with postfix values with different scale. Comments are nice, but code is better. Proper name will prevent from storing a wrong value scale in it, comment - not.

Reply:
As mentioned in https://github.com/DataDog/dd-trace-rb/pull/4028 I would like to leave the time units alone until there is team consensus on how they should be indicated and all of the other pending DI code is merged.